packages/gateway/src/manager.ts

```diff
@@ -54,6 +55,160 @@ export function createGatewayManager(options: CreateGatewayManagerOptions): Gate
       },
     },
     logger: options.logger ?? logger,
+    resharding: {
+      enabled: true,
+      shardsFullPercentage: 80,
+      checkInterval: 28800000,
+      shards: new Collection(),
+      pendingShards: new Collection(),
+      async getSessionInfo() {
+        throw new Error(
+          '[Resharding] was enabled but no getSessionInfo handler was provided. Please set a handler like: gateway.resharding.getSessionInfo = async () => { // insert code here to fetch getSessionInfo from rest process. }',
+        )
+      },
+      async checkIfReshardingIsNeeded() {
+        logger.warn(`[Resharding] Checking if resharding is needed.`)
+        // Resharding is disabled.
+        if (!gateway.resharding.enabled) return { needed: false }
+        logger.warn(`[Resharding] Resharding is enabled.`)
+
+        // TODO: fetch bot gateway info
+        const sessionInfo = await gateway.resharding.getSessionInfo()
+        logger.warn(`[Resharding] Session info retrieved.`)
+        // Don't have enough identify limits to try resharding
+        if (sessionInfo.sessionStartLimit.remaining < sessionInfo.shards) return { needed: false, info: sessionInfo }
+        logger.warn(`[Resharding] Able to reshard, checking whether necessary now.`)
+
+        // 2500 is the max amount of guilds a single shard can handle
+        // 1000 is the amount of guilds discord uses to determine how many shards to recommend.
+        // This algo helps check if your bot has grown enough to reshard.
+        const percentage = ((2500 * sessionInfo.shards) / (gateway.totalShards * 1000)) * 100
+        // Less than necessary% being used so do nothing
+        if (percentage < gateway.resharding.shardsFullPercentage) return { needed: false, info: sessionInfo }
+        logger.warn(`[Resharding] Resharding is needed.`)
+
+        return { needed: true, info: sessionInfo }
+      },
+      async reshard(info) {
+        logger.warn(`[Resharding] Starting the reshard process. Previous total shards. ${gateway.totalShards}`)
+        // Set values on gateway
+        gateway.totalShards = info.shards
+        // Handles preparing mid sized bots for LBS
```

Comment:
What's LBS?

Reply:
[Large Bot Sharding](https://discord.com/developers/docs/topics/gateway#sharding-for-large-bots) (increased `max_concurrency`). The number of shards must be a multiple of `max_concurrency`, and `gateway.calculateTotalShards()` handles that